File: Hash Table/447. Number of Boomerangs.cpp
/*
447. Number of Boomerangs


Given n points in the plane that are all pairwise distinct, a "boomerang" is a tuple of points (i, j, k) 
such that the distance between i and j equals the distance between i and k (the order of the tuple matters).

Find the number of boomerangs. You may assume that n will be at most 500 and coordinates of points are all in the range [-10000, 10000] (inclusive).

Example:
Input:
[[0,0],[1,0],[2,0]]

Output:
2

Explanation:
The two boomerangs are [[1,0],[0,0],[2,0]] and [[1,0],[2,0],[0,0]]
*/


class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int res = 0;
        for(int i = 0; i<points.size();i++){
            unordered_map<int,int>m;
            for(int j = 0; j<points.size();j++){
                if(j!=i) m[calcu(points[i], points[j])]++;
            }
            for(auto it: m)
                res += it.second*(it.second-1);
        }
        return res;
    }
    
    int calcu(pair<int,int>&a, pair<int,int>&b){
        return (a.first - b.first)*(a.first - b.first) + (a.second - b.second)*(a.second - b.second);
    }
};


class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int booms = 0;
        for (auto &p : points) {
            unordered_map<double, int> ctr(points.size());
            for (auto &q : points)
                booms += 2 * ctr[hypot(p.first - q.first, p.second - q.second)]++; //2*prev 是，新的和原来每个组 都可以成为2个新的
        }
        return booms;
    }
};

File: Hash Table/447. Number of Boomerangs.py
"""
447. Number of Boomerangs


Input:
[[0,0],[1,0],[2,0]]

Output:
2

Explanation:
The two boomerangs are [[1,0],[0,0],[2,0]] and [[1,0],[2,0],[0,0]]
"""


class Solution:
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        def caldis(a, b):
            return (a[0]-b[0])**2 + (a[1]-b[1])**2
        res = 0
        for i, v in enumerate(points):
            dic = collections.defaultdict(int)
            for j, w in enumerate(points):
                if i == j: continue
                dic[caldis(v, w)]+=1
            for v in dic.values():  
                res += (v-1)*v
        return res

class Solution:
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        ans = 0
        for p in points:
            dic = {}
            for p1 in points:
                f = p1[0] - p[0]
                s = p1[1] - p[1]
                d = f*f + s*s
                dic[d] = 1 + dic.get(d,0)
            for count in dic.values():
                ans += count*(count-1)
        return ans